### DISCUSSION TITLE
Horn Clause Predicates

### DISCUSSION THREAD

**Discussion by TheFo2sh** [dotnet/csharplang#2610] [Language Ideas]
Functional programming support in C# is getting increased , having the ability of writing horn clauses predicates simialaire to prolog in c# will be strong improvement in the language itself. 

### Suggested Syntax
` 		Even(n)=>
	                   |2 => true 
	                   |n => !Even(n-1);`

And it can be used as a predicate
`Even(4); //return true`
or with linq 
`new[]{2,3,4,5,6,7,8,9,10,11,12}.Where(Even).Sum()` 

### Related Language Feature
I made a fast prove of concept using f# pattern matching feature 
`   let rec Even  x=   match x with 
                            | 2 -> true 
                            | _ ->  not(Even(x-1))`
### Challenges
1- Infinite Recursion Problem
When the input domain contains a value out of the Predicate range , it will end up with infinite recursion and then stackoverflow. It could easily happen in the previous could by doing execution the previous code with Even(1) 
this can be solved by modifying the previous code to be

A more complete version that work with all the integers
` 		Even(n)=>
	                   |2 => true 
	                   |n =>x>2 && !Even(n-1);
                           |n =>x<2 && !Even(n+1);`

which is equivalent to

`  let rec Even  x=   match x with 
                            | 2 -> true 
                            | x when (x>2) ->  not(Even(x-1))
                            | x when (x<2) ->  not(Even(x+1))
`

2- a strong challenge to make working is the lack of tail recursion optimization in c# so any attempt to executed the previously described code will end with stackoverflow exception. 

Related proposal
https://github.com/dotnet/csharplang/issues/2544?fbclid=IwAR3ioeQ5OYA-84EczIPT7aE5gNVXX3tnRDOfF46ugGlGtU4BTWUf_9AqLyo

**Comment by YairHalberstadt** [comment]
```csharp
Even(n) => n switch { 2 => true, _ => !Even(n - 1)};
```
Is legal as of C# 8 switch expressions.

**Comment by MgSam** [comment]
Bonus points for the most over-complicated way to check if a number is even.

**Comment by The-Futurist** [comment]
@TheFo2sh 

I'll admit I've never heard of "horn clauses" despite studying functional programming for some years, this is always interesting to stumble upon a new concept. So tell me how would you express what you're trying to express in a functional language - like F#? This may help me understand it better.

A cursory search seems to show this a concept from "logic programming" rather than functional, is that true?

Also can the "infinite recursion" and consequent stack overflow be addressed by tail recursion?

**Comment by TheFo2sh** [comment]
@Korporal 
you are right horn clauses is initially a logic programming concept , as functional programming and logical programming share the same nature of being declarative Horn clauses can be used in both paradigms as a declarative constrain in other word horn clauses used to represent rules  in logic paradigm  and can be used to represent declarative condition predicate in functional paradigm.

As example of how horn clauses can be an alternative to complex conditions chain 
select social media posts that have more then 10 likes or the sum of the likes on it is recursive comments (including comments on comments) are more 30.

Normal c# code
```
bool IsValid(post)
{
        if(post.Likes.Count()>=10)
            return true;
        if(post.Comments.Select(comment=>CommentsCount(comment)).Sum()>=30)
            return true;
    return false;
}

int CommentsCount(comment){
    return comment.Likes.Count()+comment.select(subcomment=>CommentsCount(subcomment)).Sum();
}
```
Horn clauses like syntax
```
IsValid(post) =>
	            |post.Likes.Count() >=10 =>true
	            |post.Comments.Any()&&TotalLikes(post.Comments) >=30 =>true


TotalLikes(comment) =>
	            |comment.Comments.Any()	=> comment.Likes().Count()
	            |_	                    => comment.Likes.Count()+comment.comments.Select(TotalLikes).Sum()
```
While the first imperative syntax is more common and more popular , the  horn clause syntax is more declarative, we just write the conditional and not having any control on how it will get executed, imagine more business cases will emerge the horn clauses syntax will be able to get extended easier.

Regarding the f# implementation in the original  post i had 2 example implementations using the pattern matching feature in f#.

i will write a separate comment on how why the new switch expression does not solve the same cases the horn clauses like predicate could solve.

**Comment by spydacarnage** [comment]
Looks a lot like local functions to me, which I use all the time for more
declarative coding in C#
